### PR TITLE
rainerscript: add append_json() function

### PR DESCRIPTION
**Pull request by 20syldev**
### Summary
Users need to dynamically build JSON arrays and objects by adding elements without external processing or complex workarounds.

### Notes
- Tests included: `rscript_append_json.sh` and `rscript_append_json-vg.sh`
- Documentation added: `doc/source/rainerscript/functions/rs-append_json.rst`

Impact: New RainerScript function available to all users.
- **Before:** No native way to append elements to JSON arrays/objects in RainerScript.
- **After:** `append_json(array, element)` or `append_json(object, key, value)` returns a new JSON structure with the element/property added.

### CODE REVIEW

**Comment by rgerhards** [comment]
pls help me: how is this related to #6385 ?

**Comment by rgerhards** [comment]
oops, sry, my fault. The two PRs are not related, I mistook the email notificaton subject lines.

File: grammar/rainerscript.c [hunk 1]
```diff
@@ -3018,6 +3018,117 @@ static void ATTR_NONNULL() doFunct_Base64Dec(struct cnffunc *__restrict__ const
...
+            goto done_object;
+        }
+
+        struct json_object_iterator it = json_object_iter_begin(input);
+        struct json_object_iterator itEnd = json_object_iter_end(input);
+        while (!json_object_iter_equal(&it, &itEnd)) {
+            json_object_object_add(result, json_object_iter_peek_name(&it),
+                                   json_object_get(json_object_iter_peek_value(&it)));
+            json_object_iter_next(&it);
+        }
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c]: rsyslog uses libfastjson where `json_object_object_add` returns `void`, so we can't check its return value. The NULL check for `newVal` has been added.

File: grammar/rainerscript.c [hunk 2]
```diff
@@ -3018,6 +3018,117 @@ static void ATTR_NONNULL() doFunct_Base64Dec(struct cnffunc *__restrict__ const
...
+        } else if (srcVal[2].datatype == 'S') {
+            char *str = (char *)var2CString(&srcVal[2], &bMustFree2);
+            if (str != NULL) {
+                newVal = json_object_new_string(str);
+            }
+            if (bMustFree2) free(str);
+        } else {
+            newVal = json_object_new_int64(srcVal[2].d.n);
+        }
+        json_object_object_add(result, key, newVal);
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c]: rsyslog uses libfastjson where `json_object_object_add` returns `void`, so we can't check its return value. The NULL check for `newVal` has been added.

**Comment by rgerhards** [comment]
@20syldev looks like this needs a rebase (not merge)

**Comment by 20syldev** [comment]
> @20syldev looks like this needs a rebase (not merge)

Done!

File: grammar/rainerscript.c [hunk 3]
```diff
@@ -3104,6 +3104,119 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
     varFreeMembers(&srcVal[1]);
 }
 
+static void ATTR_NONNULL() doFunct_append_json(struct cnffunc *__restrict__ const func,
+                                               struct svar *__restrict__ const ret,
+                                               void *const usrptr,
+                                               wti_t *const pWti) {
+    struct svar srcVal[3];
+    int bMustFree = 0;
+    int bMustFree2 = 0;
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c:3265]: `bMustFree` is a common naming pattern used throughout the codebase. Kept for consistency.

File: grammar/rainerscript.c [hunk 4]
```diff
@@ -3104,6 +3104,119 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+    done_object:
+        if (bMustFree) free(key);
+        varFreeMembers(&srcVal[1]);
+        varFreeMembers(&srcVal[2]);
+    }
+
+    ret->datatype = 'J';
+    ret->d.json = result;
+    varFreeMembers(&srcVal[0]);
+}
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c:3349]: Fixed. Refactored to single `finalize_it`: label with `nParamsEvaluated` counter.

File: grammar/rainerscript.c [hunk 5]
```diff
@@ -3104,6 +3104,119 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+        }
+        if (newVal != NULL) {
+            json_object_object_add(result, key, newVal);
+        }
+
+    done_object:
+        if (bMustFree) free(key);
+        varFreeMembers(&srcVal[1]);
+        varFreeMembers(&srcVal[2]);
+    }
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c:3338]: libfastjson's `json_object_object_add_ex` returns `void`, not `int`. Error checking not available with this API.

File: tests/Makefile.am [hunk 6]
```diff
@@ -626,6 +627,7 @@ TESTS +=  \
 	func-substring-invld-startpos-vg.sh \
 	rscript_trim-vg.sh \
 	rscript_split-vg.sh
```
Comment (rgerhards) [inline tests/Makefile.am]: There is the coninuation line indicator \ missing - thus the build failue.

File: tests/rscript_append_json-vg.sh [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
```
Comment (rgerhards) [inline tests/rscript_append_json-vg.sh:1]: this looks like the same test as the non-valgrind version. check `zstd-vg.sh` to see how this can be done better without copying code.

Comment (20syldev) [reply tests/rscript_append_json-vg.sh:1]: Thanks ! I'll check that out

File: tests/Makefile.am (hunk 6 above)
Comment (20syldev) [reply tests/Makefile.am]: Same mistake as in my previous pull request ah !

File: tests/rscript_append_json-vg.sh (hunk 7 above)
Comment (20syldev) [reply tests/rscript_append_json-vg.sh:1]: > this looks like the same test as the non-valgrind version. check `zstd-vg.sh` to see how this can be done better without copying code.

I could also make this change to `rscript_split-vg.sh`...

Comment (20syldev) [reply tests/rscript_append_json-vg.sh:1]: Do you think it would be better to use source `source ${srcdir:-$(dirname “$0”)}/<script>` instead of `source ${srcdir:-.}/<script>` so that the test can be run from anywhere? (e.g., the project root)

Comment (rgerhards) [reply tests/rscript_append_json-vg.sh:1]: that's probably a good idea!

File: grammar/rainerscript.c [hunk 8]
```diff
@@ -3104,6 +3104,122 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+        for (int i = 0; i < len; i++) {
+            struct json_object *elem = json_object_array_get_idx(input, i);
+            struct json_object *elem_ref = json_object_get(elem);
+            if (json_object_array_add(result, elem_ref) != 0) {
+                json_object_put(elem_ref);
+                json_object_put(result);
+                result = NULL;
+                goto finalize_it;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c]: rsyslog uses `libfastjson`, not `json-c`, and `json_object_deep_copy()` is not available in this library.

File: grammar/rainerscript.c [hunk 9]
```diff
@@ -3104,6 +3104,122 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+            goto finalize_it;
+        }
+
+        struct json_object_iterator it = json_object_iter_begin(input);
+        struct json_object_iterator itEnd = json_object_iter_end(input);
+        while (!json_object_iter_equal(&it, &itEnd)) {
+            json_object_object_add(result, json_object_iter_peek_name(&it),
+                                   json_object_get(json_object_iter_peek_value(&it)));
+            json_object_iter_next(&it);
+        }
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c]: Same as above - `json_object_deep_copy()` is not available in `libfastjson`.

File: grammar/rainerscript.c [hunk 10]
```diff
@@ -3104,6 +3104,122 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+            newElem = json_object_get(srcVal[1].d.json);
+        } else if (srcVal[1].datatype == 'S') {
+            char *str = (char *)var2CString(&srcVal[1], &bMustFree2);
+            if (str != NULL) {
+                newElem = json_object_new_string(str);
+            }
+            if (bMustFree2) free(str);
+        } else {
+            newElem = json_object_new_int64(srcVal[1].d.n);
+        }
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c]: Fixed

File: grammar/rainerscript.c [hunk 11]
```diff
@@ -3104,6 +3104,122 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+            newVal = json_object_get(srcVal[2].d.json);
+        } else if (srcVal[2].datatype == 'S') {
+            char *str = (char *)var2CString(&srcVal[2], &bMustFree2);
+            if (str != NULL) {
+                newVal = json_object_new_string(str);
+            }
+            if (bMustFree2) free(str);
+        } else {
+            newVal = json_object_new_int64(srcVal[2].d.n);
+        }
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c]: Fixed

**Comment by 20syldev** [comment]
Please let me know if any of the issues I marked as “Resolved” are possible, as I haven't figured out how to do it with libfastjson.

File: grammar/rainerscript.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (rgerhards) [reply grammar/rainerscript.c]: This looks very valid, given the overall design. Note that there is a function `jsonDeepCopy ()` in msg.c.

File: grammar/rainerscript.c [hunk 12]
```diff
@@ -3104,6 +3104,122 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+            goto finalize_it;
+        }
+
+        struct json_object_iterator it = json_object_iter_begin(input);
+        struct json_object_iterator itEnd = json_object_iter_end(input);
+        while (!json_object_iter_equal(&it, &itEnd)) {
+            json_object_object_add(result, json_object_iter_peek_name(&it),
+                                   json_object_get(json_object_iter_peek_value(&it)));
+            json_object_iter_next(&it);
+        }
```
[the remark this answers is not in the document]
Comment (rgerhards) [reply grammar/rainerscript.c]: same as above

File: grammar/rainerscript.c [hunk 13]
```diff
@@ -3104,6 +3104,122 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+            if (str != NULL) {
+                newVal = json_object_new_string(str);
+            }
+            if (bMustFree2) free(str);
+        } else if (srcVal[2].datatype == 'N') {
+            newVal = json_object_new_int64(srcVal[2].d.n);
+        }
+        if (newVal != NULL) {
+            json_object_object_add(result, key, newVal);
+        }
```
[the remark this answers is not in the document]
Comment (rgerhards) [reply grammar/rainerscript.c:3337]: i did not check if this was newly interoduced, but refactoring would be good in any case.

[the remark this answers is not in the document]
Comment (rgerhards) [reply grammar/rainerscript.c:3337]: can be ignored

File: tests/rscript_append_json.sh [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+    action(type="omfile" file=`echo $RSYSLOG_OUT_LOG` template="outfmt")
+}
+'
+startup
+tcpflood -m1
+shutdown_when_empty
+wait_shutdown
+
+export EXPECTED='[ "a", "b", "c" ]|{ "name": "test", "status": "active" }|[ "first", "second" ]'
```
[the remark this answers is not in the document]
Comment (rgerhards) [reply tests/rscript_append_json.sh:33]: this is sufficient for our use case

File: grammar/rainerscript.c [hunk 15]
```diff
@@ -3104,6 +3104,102 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+    }
+
+    input = srcVal[0].d.json;
+    type = json_object_get_type(input);
+
+    if (type == json_type_array) {
+        cnfexprEval(func->expr[1], &srcVal[1], usrptr, pWti);
+        nParamsEvaluated = 1;
+
+        result = jsonDeepCopy(input);
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c:3283]: `jsonDeepCopy()` is an internal rsyslog function from `msg.c`. This is the function rgerhards recommended in the previous review.

File: grammar/rainerscript.c [hunk 16]
```diff
@@ -3104,6 +3104,102 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+            if (str != NULL) {
+                newVal = json_object_new_string(str);
+            }
+            if (bMustFree2) free(str);
+        } else if (srcVal[2].datatype == 'N') {
+            newVal = json_object_new_int64(srcVal[2].d.n);
+        }
+        if (newVal != NULL) {
+            json_object_object_add(result, key, newVal);
+        }
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c:3337]: rgerhards already addressed this in the previous review.

File: grammar/rainerscript.c [hunk 17]
```diff
@@ -3104,6 +3104,102 @@ static void ATTR_NONNULL() doFunct_split(struct cnffunc *__restrict__ const func
...
+            newElem = jsonDeepCopy(srcVal[1].d.json);
+        } else if (srcVal[1].datatype == 'S') {
+            char *str = (char *)var2CString(&srcVal[1], &bMustFree2);
+            if (str != NULL) {
+                newElem = json_object_new_string(str);
+            }
+            if (bMustFree2) free(str);
+        } else if (srcVal[1].datatype == 'N') {
+            newElem = json_object_new_int64(srcVal[1].d.n);
+        }
```
[the remark this answers is not in the document]
Comment (20syldev) [reply grammar/rainerscript.c:3299]: Keeping the code inline as requested.

**Comment by rgerhards** [comment]
@20syldev my bad, I had a wrong merge order. I'll see that I can rebase it.

**Comment by 20syldev** [comment]
> @20syldev my bad, I had a wrong merge order. I'll see that I can rebase it.

No problem

**Review by rgerhards — APPROVED** [review]

**Comment by rgerhards** [comment]
CI fail is a know and unrelated flake

**Comment by 20syldev** [comment]
> CI fail is a know and unrelated flake

Great, I was wondering about that. Thanks for the merge!

**Comment by 20syldev** [comment]
@rgerhards Hi, I just noticed that the commit message was copied incorrectly for append_json. Here is the correct one: 

Title:
```
rainerscript: add append_json() function
```
Description:
```
Users need to dynamically build JSON arrays and objects by adding
elements without external processing or complex workarounds.

Impact: New RainerScript function available to all users.

Before: No native way to append elements to JSON arrays/objects in RainerScript.
After: append_json(array, element) or append_json(object, key, value) returns
a new JSON structure with the element/property added.

Technical overview:
Implements doFunct_append() in grammar/rainerscript.c
Registers "append_json" in scriptFunct table with 2-3 args
Adds CNFFUNC_APPEND_JSON enum in rainerscript.h
Supports both JSON arrays (2 args) and objects (3 args)
Returns new object (functional style, original unchanged)
Uses json_object_iter API for object copying
Handles edge cases: null input, non-JSON input, memory allocation failures
Includes documentation (rs-append_json.rst) and test scripts
```

**Comment by rgerhards** [comment]
sry for that, but now it's too late. It is impossible to change commit afterwards.

**Comment by 20syldev** [comment]
> sry for that, but now it's too late. It is impossible to change commit afterwards.

Okay, I understand!
